### PR TITLE
Add methods related to node options to NodeBuilder

### LINKED ISSUE #142: Add support for node options
Perhaps not all options should be exposed yet (e.g. we don't have rosout logging yet), but something like `use_global_arguments` could definitely be useful.

### PR DESCRIPTION
**Pull request by Soya-Onishi**
Closes #142 

This PR add node options related to [`rcl_node_options_t`](https://docs.ros2.org/latest/api/rcl/structrcl__node__options__t.html).
However, methods related to allocator options (e.g. NodeBuilder::allocators.allocate) are not implemented because there are no way to return allocated heap pointer as far as I know.

### CODE REVIEW

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:303]:
```suggestion
        // SAFETY: No preconditions for this function.
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:275]: Can you move this line further down, close to `rcl_node_init()` where it is used?

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:47]: The domain ID isn't part of the node_options anymore from galactic onwards. I think it now is an attribute of the context, not the node: https://github.com/ros2/rcl/pull/708

So, probably better to remove this. Sorry, I think I explicitly suggested this on the issue!

**Review by nnmm — COMMENTED** [review]
Thanks! I'll add more comments later.

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs:47]: Or, instead of removing, you can gate it with `#[cfg(ros_distro = "foxy")]`.

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:50]: Thanks for your suggestions! I have one question. If making fields public, should context filed also be public? Current context field has raw type of rcl and should not expose those types against users as much as possible I think. If it should make context field private, it is better for other fields to make private because of the rule, `Either make all fields of a type public, or make none of them public`, from blog post that is linked from reddit you attached.  Another way to make all field public is usage of reference for Context, but this may make compiler more sensitive and cause lifetime errors which is hard to understand why, although I haven't inspect yet.

By the way, surely, current builder setters are smell ones if I do not misunderstand the means about smell( I read posts as smell setter is equal to trivial setter, and trivial setter is the setter that takes a parameter and only assigns to struct field). However I could not understand why setter is worse when there are smell setters. If there are comments says why smell setters are bad, please tell me.

Also, the post you attached does not says about assigning a value to public field as far as I know. I read that post as recommendation of using C-like struct instantiation pattern with Default when the struct just be data box. However, on another note, this pattern is possible only if there are only optional values (there is also same questionable comment about that but no response), and NodeBuilder has essential fields (i.e. `context` and `name`), so, in this case, I cannot apply this pattern to NodeBuilder.

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:47]: Thank you for letting me know about `domain_id`.
Please tell me where the description or documentation about `domain_id` is in context now.

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs:50]: Sorry, I guess that link is a bit confusing without context. Let me try to clarify
- I agree, the context should definitely be private.
- The "trivial setters" matklad calls "smelly" are those in the parent comment: `shape.set_position(Vec3(0, 9, 2));` etc. – they have a signature like `fn set_position(&mut self, position: Vec3)`.
  - This signature is different from the chainable setters in `NodeBuilder`, which look more like `fn position(self, position: Vec3) -> Self`. Chainable setters are useful and I think no one sees them as a code smell. So I think there's nothing wrong at all with the current implementation of `NodeBuilder`.
- So, the point of linking to that comment was just that public fields are not always bad. The "builder lite" blog post it's attached to isn't really relevant here. Sorry for the confusion.
- Further down in the reddit thread, people agree that public fields are fine unless (1) there are some invariants that need to be protected (we don't have that, since validation happens later) or (2) you need to be careful about making breaking changes to the struct (which is also not a concern, since we can make breaking changes, and since the private context field will make the changes non-breaking anyway).
- Yep, https://matklad.github.io/2022/05/29/binary-privacy.html recommends categorizing everything into data or abstract data types. Not sure which one a builder is supposed to be. I think it's closer to data, since we could also remove the context field and only pass it in in the `build()` method, then everything could be public. But in the end it's only a rule of thumb, just like "avoid trivial setters".

It's a really small change we're discussing here, so I'll leave it up to you :) I just find it nice to see the fields in the docs and see that there is nothing weird going on, and to be able to also use `builder.enable_rosout = true;` syntax.

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs:47]: In newer distros it's part of the `init_options` that are passed to `rcl_init()` (called in the `Context::new` method).

Comment (nnmm) [reply rclrs/src/node/builder.rs:47]: Probably it's best to leave the `domain_id` for a separate PR.

**Comment by nnmm** [comment]
@Soya-Onishi Great work! Could you also add `Drop` instances for `rcl_node_options_t` and `rcl_arguments_t` that call their `fini` functions?

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:283]:
```suggestion
            // SAFETY: This function does not store the ephemeral cstring_args_ptrs
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:16]:
```suggestion
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:19]:
```suggestion
/// - `enable_rosout: true`
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:210]: I think it's good to just link to `Context::new()` (which refers to the official ROS 2 article), to have one source of truth for how command line arguments for ROS 2 work.

```suggestion
    /// Sets node-specific command line arguments.
    ///
    /// These arguments are parsed the same way as those for [`Context::new()`][1].
    /// However, the node-specific command line arguments have higher precedence than the arguments
    /// used in creating the context.
    ///    
    /// # Example
    /// ```
    /// # use rclrs::{Context, Node, NodeBuilder, RclrsError};
    /// // Usually, this would change the name of "my_node" to "context_args_node":
    /// let context_args = ["--ros-args", "--remap", "my_node:__node:=context_args_node"]
    ///   .map(String::from);
    /// let context = Context::new(context_args)?;
    /// // But the node arguments will change it to "node_args_node":
    /// let node_args = ["--ros-args", "--remap", "my_node:__node:=node_args_node"]
    ///   .map(String::from);
    /// let node = context
    ///   .create_node_builder("my_node")
    ///   .arguments(node_args)
    ///   .build()?;
    /// assert_eq!(node.name(), "node_args_node");
    /// # Ok::<(), RclrsError>(())
    /// ```
    ///
    /// [1]: crate::Context::new
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:218]: Do you know if this actually affects logs emitted by the `rcl` layer? Or does it currently not have any effect at all?

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:180]:
```suggestion
    /// Enables or disables using global arguments.
    ///
    /// The "global" arguments are those used in [creating the context][1].
    ///
    /// # Example
    /// ```
    /// # use rclrs::{Context, Node, NodeBuilder, RclrsError};
    /// let context_args = ["--ros-args", "--remap", "__node:=your_node"]
    ///   .map(String::from);
    /// let context = Context::new(context_args)?;
    /// // Ignore the global arguments:
    /// let node_without_global_args = context
    ///   .create_node_builder("my_node")
    ///   .use_global_arguments(false)
    ///   .build()?;
    /// assert_eq!(node_without_global_args.name(), "my_node");
    /// // Do not ignore the global arguments:
    /// let node_with_global_args = context
    ///   .create_node_builder("my_other_node")
    ///   .use_global_arguments(true)
    ///   .build()?;
    /// assert_eq!(node_with_global_args.name(), "your_node");
    /// # Ok::<(), RclrsError>(())
    /// ```
    ///
    /// [1]: crate::Context::new
```

**Comment by nnmm** [comment]
I tested it out and got an error about `rcl_arguments_fini` being called twice (which makes sense, because `rcl_node_options_fini` also calls `rcl_arguments_fini`). So we should probably remove the `Drop` impl for `rcl_arguments_t` again, that would be easiest.

Apart from that, I only have one other comment (about enable_rosout), then it's ready to merge!

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:219]:
```suggestion
    /// Enables or disables logging to rosout.
    ///
    /// When enabled, log messages are published to the `/rosout` topic in addition to
    /// standard output.
    ///
    /// This option is currently unused in `rclrs`.
```

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs:220]: I believe it currently has no effect.

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:235]:
```suggestion
    /// If any of the below conditions are fulfilled, a panic occurs:
    ///  - The node name contains a null byte
    ///  - The namespace contains a null byte
    ///  - Any of the node arguments contain a null byte
```

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:220]: As for now, there is no effect in ros2_rust.
However, as far as I know, this flag affects whether logging features publish to `rosout` topic or not.
For the future, this should retain and put comments to tell this is nothing to do for now, I think.

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:219]: Thanks for adding more detailed comments!!
As I said above, this will work after implementing logging features in rclrs I think.

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:285]: This line is outdated and can be removed.

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:307]: I think it would be slightly cleaner to just remove this impl since it will always be freed as part of `rcl_node_options_fini`. But if you disagree, feel free to leave it in.

Comment (esteve) [reply rclrs/src/node/builder.rs:307]: I can't think of an example where `rcl_arguments_t` would be used directly and not in combination with `rcl_node_options_t`, except for parsing command line arguments, but we'd use Rust's native facilities in that case. I can understand for the sake of completeness, it's up to you @Soya-Onishi if you want to keep this code, I'm leaning towards removing it.

**Review by esteve — CHANGES_REQUESTED** [review]

**Comment by nnmm** [comment]
@Soya-Onishi Could you adjust the CString operations to match https://github.com/ros2-rust/ros2_rust/pull/190, which was merged in the meantime?

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs:47]: Feel free to resolve this, it was just a suggestion and like I said I'll leave it up to you.

**Comment by nnmm** [comment]
@Soya-Onishi I have another PR that conflicts with this one, https://github.com/ros2-rust/ros2_rust/pull/193. If you plan to make the couple of remaining fixes this week, I can wait with my PR until we merge this one, so you don't have extra work.

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:307]: @esteve Thanks for your advice! I don't know whether it is related to what @nnmm says or not, but should I put statements about `rcl_arguments_t` is finished in `rcl_node_options_fini()`?

Comment (nnmm) [reply rclrs/src/node/builder.rs:307]: Yes, we're both saying the same :) You could add a comment `// This also finalizes the rcl_arguments contained in the node options`  to the `Drop for rcl_node_options_t`, and then just delete the `Drop for rcl_arguments_t`.

File: rclrs/src/node/builder.rs
Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs:285]: Thanks for your pick! I'll fix it.

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:313]:
```suggestion
            // This also finalizes the `rcl_arguments_t` contained in `rcl_node_options_t`.
```

Super nitpicky, I know :)

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:276]:
```suggestion
    ///
    /// Any fields not present in the builder will have their default value.
```

**Review by esteve — DISMISSED** [review]
@Soya-Onishi thanks for all the hard work and for the patience iterating with us! :slightly_smiling_face:

File: rclrs/src/node/builder.rs
Comment (nnmm) [inline rclrs/src/node/builder.rs:296]: Did you know you can collect into a `Result<Vec<_>>`? That would be better here, I think.

Comment (Soya-Onishi) [reply rclrs/src/node/builder.rs]: Thanks for your notice!

File: rclrs/src/node/builder.rs
Comment (nnmm) [reply rclrs/src/node/builder.rs]: I just removed it since it was the only remaining fix, hope that's okay!

**Review by nnmm — APPROVED** [review]

**Comment by nnmm** [comment]
Wonderful! Thanks @Soya-Onishi!

**Comment by Soya-Onishi** [comment]
@nnmm @esteve 
Thanks for your many advices and picks!

**Comment by nnmm** [comment]
Feel free to join the Matrix chat by the way.

**Pull request merged**
